src/Compiler/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -1223,6 +1223,191 @@ public class ComponentWithEditorRequiredChildContent : ComponentBase
+            }
+            """));
+
+        var generated = CompileToCSharp("""
+            <MyComponent @bind-MyParameter="c"
+                BoolParameter="true"
+                StringParameter="str"
+                DelegateParameter="() => { }"
+                ObjectParameter="c" />
+
+            @code {
+                private MyClass<string> c = new();
+            }
+            """);
+
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated);
+    }
+
+    [Fact] // https://github.com/dotnet/aspnetcore/issues/18042
+    public void AddAttribute_ImplicitStringConversion_CustomEvent()
+    {
+        AdditionalSyntaxTrees.Add(Parse("""
+            using Microsoft.AspNetCore.Components;
+
+            namespace Test;
+            
+            public class MyClass<T>
+            {
+                public static implicit operator string(MyClass<T> c) => throw null!;
+            }
+
+            public class MyComponent<T> : ComponentBase
+            {
+                [Parameter]
+                public MyClass<T> MyParameter { get; set; }
+
+                [Parameter]
+                public EventCallback MyEvent { get; set; }
+            
+                [Parameter]
+                public bool BoolParameter { get; set; }
+            
+                [Parameter]
+                public string StringParameter { get; set; } = null!;
+            
+                [Parameter]
+                public System.Delegate DelegateParameter { get; set; } = null!;
+            
+                [Parameter]
+                public object ObjectParameter { get; set; } = null!;
+            }
+            """));
+
+        var generated = CompileToCSharp("""
+            <MyComponent MyParameter="c"
+                MyEvent="() => { }"
+                BoolParameter="true"
+                StringParameter="str"
+                DelegateParameter="() => { }"
+                ObjectParameter="c" />
+
+            @code {
+                private MyClass<string> c = new();
+            }
+            """);
+
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated);
+    }
+
+    [Fact] // https://github.com/dotnet/aspnetcore/issues/18042
+    public void AddAttribute_ImplicitStringConversion_BindUnknown()
+    {
+        AdditionalSyntaxTrees.Add(Parse("""
+            using Microsoft.AspNetCore.Components;
+
+            namespace Test;
+            
+            public class MyClass
+            {
+                public static implicit operator string(MyClass c) => throw null!;
+            }
+
+            public class MyComponent : ComponentBase
+            {
+            }
+            """));
+
+        var generated = CompileToCSharp("""
+            <MyComponent @bind-Value="c" />
```

Comment:
Let's have a version of this test that will actually fail if the parens are missing from the `AddComponent` call.

Reply:
Added.